drizzle-orm/type-tests/pg/set-operators.ts

```diff
@@ -0,0 +1,194 @@
+				.from(users),
+			db
+				.select({ id: users.id, homeCity: sql`${users.homeCity}`.mapWith(Number) })
+				.from(users),
+		)
+	);
+
+Expect<Equal<{ id: number; homeCity: number }[], typeof intersectTest>>;
+
+const intersectAllTest = await db
+	.select({ id: users.id, homeCity: users.class })
+	.from(users)
+	.intersectAll(
+		db
+			.select({ id: users.id, homeCity: users.class })
+			.from(users)
+			.leftJoin(cities, eq(users.id, cities.id)),
+	);
+
+Expect<Equal<{ id: number; homeCity: 'A' | 'C' }[], typeof intersectAllTest>>;
+
+const exceptTest = await db
+	.select({ id: users.id, homeCity: users.homeCity })
+	.from(users)
+	.except(
+		db
+			.select({ id: users.id, homeCity: sql`${users.homeCity}`.mapWith(Number) })
+			.from(users),
+	);
+
+Expect<Equal<{ id: number; homeCity: number }[], typeof exceptTest>>;
+
+const exceptAllTest = await db
+	.select({ id: users.id, homeCity: users.class })
+	.from(users)
+	.exceptAll(
+		db
+			.select({ id: users.id, homeCity: sql<'A' | 'C'>`${users.class}` })
+			.from(users),
+	);
+
+Expect<Equal<{ id: number; homeCity: 'A' | 'C' }[], typeof exceptAllTest>>;
+
+const union2Test = await union(db.select().from(cities), db.select().from(cities), db.select().from(cities));
+
+Expect<Equal<{ id: number; name: string; population: number | null }[], typeof union2Test>>;
+
+const unionAll2Test = await unionAll(
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+	db.select().from(cities),
+);
+
+Expect<Equal<{ id: number; name: string; population: number | null }[], typeof unionAll2Test>>;
+
+const intersect2Test = await intersect(
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+);
+
+Expect<Equal<{ id: number; name: string; population: number | null }[], typeof intersect2Test>>;
+
+const intersectAll2Test = await intersectAll(
+	db.select({
+		id: cities.id,
+	}).from(cities),
+	db.select({
+		id: cities.id,
+	})
+		.from(cities),
+).orderBy(desc(cities.id)).limit(23);
+
+Expect<Equal<{ id: number }[], typeof intersectAll2Test>>;
+
+const except2Test = await except(
+	db.select({
+		userId: newYorkers.userId,
+	})
+		.from(newYorkers),
+	db.select({
+		userId: newYorkers.userId,
+	}).from(newYorkers),
+);
+
+Expect<Equal<{ userId: number }[], typeof except2Test>>;
+
+const exceptAll2Test = await exceptAll(
+	db.select({
+		userId: newYorkers.userId,
+		cityId: newYorkers.cityId,
+	})
+		.from(newYorkers),
+	db.select({
+		userId: newYorkers.userId,
+		cityId: newYorkers.cityId,
+	}).from(newYorkers),
+);
+
+Expect<Equal<{ userId: number; cityId: number | null }[], typeof exceptAll2Test>>;
+
+// @ts-expect-error - The select on both sites must be the same shape
```

Comment:
Maybe let's change an error to this. I took it from the Postgres docs, as it seems to be more specific about what was wrong

> All queries in combining statements should return the same number of columns and the corresponding columns should have compatible data type